extension/encoding/awslogsencodingextension/internal/unmarshaler/cloudtraillogs/cloudtraillogs.go

```diff
@@ -0,0 +1,283 @@
+}
+
+var _ plog.Unmarshaler = (*CloudTrailLogsUnmarshaler)(nil)
+
+// CloudTrailRecord represents a CloudTrail log record
+// There is no builtin CloudTrailRecord we can leverage like in S3
+// So we build our own
+type CloudTrailRecord struct {
+	EventVersion                 string         `json:"eventVersion"`
+	EventTime                    string         `json:"eventTime"`
+	EventSource                  string         `json:"eventSource"`
+	EventName                    string         `json:"eventName"`
+	AwsRegion                    string         `json:"awsRegion"`
+	SourceIPAddress              string         `json:"sourceIPAddress"`
+	UserAgent                    string         `json:"userAgent"`
+	RequestID                    string         `json:"requestID"`
+	EventID                      string         `json:"eventID"`
+	EventType                    string         `json:"eventType"`
+	EventCategory                string         `json:"eventCategory"`
+	RecipientAccountID           string         `json:"recipientAccountId"`
+	UserIdentity                 map[string]any `json:"userIdentity"`
+	ResponseElements             map[string]any `json:"responseElements"`
+	RequestParameters            map[string]any `json:"requestParameters"`
+	Resources                    []any          `json:"resources"`
+	ReadOnly                     bool           `json:"readOnly"`
+	ManagementEvent              bool           `json:"managementEvent"`
+	AdditionalEventData          map[string]any `json:"additionalEventData"`
+	TLSDetails                   map[string]any `json:"tlsDetails"`
+	SessionCredentialFromConsole string         `json:"sessionCredentialFromConsole"`
+}
+
+func NewCloudTrailLogsUnmarshaler(buildInfo component.BuildInfo) plog.Unmarshaler {
+	return &CloudTrailLogsUnmarshaler{
+		buildInfo: buildInfo,
+		gzipPool:  sync.Pool{},
+	}
+}
+
+func (u *CloudTrailLogsUnmarshaler) Unmarshal(buf []byte) (plog.Logs, error) {
+	return u.UnmarshalLogs(buf)
+}
+
+func (u *CloudTrailLogsUnmarshaler) UnmarshalLogs(buf []byte) (plog.Logs, error) {
+	// Check if the content is gzip-compressed
+	var errDecompress error
+	gzipReader, ok := u.gzipPool.Get().(*gzip.Reader)
+	if !ok {
+		gzipReader, errDecompress = gzip.NewReader(bytes.NewReader(buf))
+	} else {
+		errDecompress = gzipReader.Reset(bytes.NewReader(buf))
+	}
+
+	var decompressedBuf []byte
+	if errDecompress == nil {
+		// Content is gzip-compressed, decompress it
+		defer func() {
+			_ = gzipReader.Close()
+			u.gzipPool.Put(gzipReader)
+		}()
+
+		decompressedBuf, errDecompress = io.ReadAll(gzipReader)
+		if errDecompress != nil {
+			return plog.Logs{}, fmt.Errorf("failed to decompress CloudTrail logs: %w", errDecompress)
+		}
+	} else {
+		// Content is not gzip-compressed, use the original buffer
+		decompressedBuf = buf
+		if gzipReader != nil {
+			u.gzipPool.Put(gzipReader)
+		}
+	}
+
+	var record CloudTrailRecord
+	if err := json.Unmarshal(decompressedBuf, &record); err != nil {
```

Comment:
Does the `decompressedBuf` contain only one cloudtrail record? Just double-checking since other JSON compressed logs are usually split by new line (so each line would be a record)

Reply:
switched the implementation to read the {"Records": [...]} object and process all the records, grouping records by resource (cloud id/cloud region)